mautic/mautic
# Microsoft Dynamics FetchLeads  - update

[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | 
| New feature? | y
| Automated tests included? |
| Related user documentation PR URL | tba
| Related developer documentation PR URL | 
| Issues addressed (#s or URLs) | 
| BC breaks? | 
| Deprecations? | 

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

This bugfix/feature has been sponsored by @Webmecanik

[//]: # ( Required: )
#### Description:
Follow up to https://github.com/mautic/mautic/pull/5229 (@npracht  should close old PR)

This PR add the ability for users of Microsoft Dynamics users
- fetch leads from leads Microsoft Dynamics object (contacts and companies are supported so far)
- resolve duplicate contact if contact exists on both sides Mautic and Microsoft Dynamics
  -  we're trying find existed contact and if exists, we just create integration entity and update contact fields based on plugin settings

![image](https://user-images.githubusercontent.com/462477/35872715-d6847f3c-0b67-11e8-83c7-3b4338ca639e.png)

#### Steps to test this PR:

1. Steps to test this PR:
2. Connect your integration (follow documentation)
3. In the feature tab, enable Leads
4. Create a lead in you MSD account
5. Launch command
6. See your lead in mautic contacts
7. Create contacts on both sides Mautic and Microsoft Dynamics
8. Try setup plugin, especially new option Mautic contacts pushed to Dynamics CRM object and Try find exists contact during the push to Dynamics CRM (see image example above)
9. Test if works properly 

<a href="https://gitpod.io/#https://github.com/mautic/mautic/pull/5674"><img src="https://gitpod.io/button/open-in-gitpod.svg"/></a>
